# Data Collection

In [1]:
# Import Basic Packages
import requests
import time
import random
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
pd.set_option('display.max_columns', 999)

## Making a Reddit Scraping Function

Firstly, let's try to scrape data from a Reddit URL and see what are the key actions we need to perform a successful web scraping. The `requests` library is used for submitting HTTP requests from Python. Let's first try it on the software engineering subreddit URL link.

In [2]:
url = 'https://www.reddit.com/r/softwareengineering.json'
headers = {'User-agent': 'GA boy'}
res = requests.get(url, headers=headers)
res.status_code

200

A HTTP Response Status Code of `200` means that the request has successed. In this example, since we have used the `GET` method, this implies that the resource has been fetched and is transmitted in the message body.

In [3]:
# Converting the data into a format similar to a dictionary using the .json() method
url_dict = res.json()
url_dict

{'kind': 'Listing',
 'data': {'modhash': '',
  'dist': 26,
  'children': [{'kind': 't3',
    'data': {'approved_at_utc': None,
     'subreddit': 'SoftwareEngineering',
     'selftext': "Hi /r/SoftwareEngineering - This is just to let you all know that automoderator has been enabled for this subreddit and we will be fine-tuning the options as we go, if you get a post filtered and you think it's an error please contact me so that I can take a look.\n\n&amp;nbsp;\n\nI also enabled a [dark theme](https://www.reddit.com/r/darkserene/) for the people not using RES' nightmode already that I hope it's easier on your eyes.\n\n&amp;nbsp;\n\nPlease use this space for any feedback you have for the community.\n\n&amp;nbsp;\n\n\nThat would be all!\n\n&amp;nbsp;\n\nAll the best,\n\nThe modteam.",
     'author_fullname': 't2_6w5wq',
     'saved': False,
     'mod_reason_title': None,
     'gilded': 0,
     'clicked': False,
     'title': 'Style changes and automoderator',
     'link_flair_richtext': [

Based on the output above, we would like to exrract columns like 'children' and further sub-columns like 'subreddit', 'selftext' and 'title' that contains key words text that we will use for our analysis.

In [4]:
url_dict['data'].keys()

dict_keys(['modhash', 'dist', 'children', 'after', 'before'])

In [5]:
# Shows the subreddit topic for the first post
url_dict['data']['children'][0]['data']['subreddit']

'SoftwareEngineering'

In [6]:
# Shows the title of the second post
url_dict['data']['children'][1]['data']['title']

'How do your companies keep track of multiple systems, products and teams to better inform decision making?'

In [7]:
# Shows the selftext of the first post
url_dict['data']['children'][0]['data']['selftext']

"Hi /r/SoftwareEngineering - This is just to let you all know that automoderator has been enabled for this subreddit and we will be fine-tuning the options as we go, if you get a post filtered and you think it's an error please contact me so that I can take a look.\n\n&amp;nbsp;\n\nI also enabled a [dark theme](https://www.reddit.com/r/darkserene/) for the people not using RES' nightmode already that I hope it's easier on your eyes.\n\n&amp;nbsp;\n\nPlease use this space for any feedback you have for the community.\n\n&amp;nbsp;\n\n\nThat would be all!\n\n&amp;nbsp;\n\nAll the best,\n\nThe modteam."

In [8]:
# Convert these subreddit posts into a pandas dataframe
url_posts = [p['data'] for p in url_dict['data']['children']]
print(str('The number of subreddit posts in this dataframe is: '), len(pd.DataFrame(url_posts)))
pd.DataFrame(url_posts).head()

The number of subreddit posts in this dataframe is:  26


approved_at_utc            subreddit  \
0            None  SoftwareEngineering   
1            None  SoftwareEngineering   
2            None  SoftwareEngineering   
3            None  SoftwareEngineering   
4            None  SoftwareEngineering   

                                            selftext author_fullname  saved  \
0  Hi /r/SoftwareEngineering - This is just to le...        t2_6w5wq  False   
1  I have been at several large companies now and...      t2_59dwm06  False   
2  I am doing a Uni course on Data Engineering, w...     t2_57j4x5fp  False   
3                                                           t2_ayl6m  False   
4  I’ve been learning to code using Udemy and thi...     t2_60gqusa6  False   

  mod_reason_title  gilded  clicked  \
0             None       0    False   
1             None       0    False   
2             None       0    False   
3             None       0    False   
4             None       0    False   

                                               title link_flair_richtext  \
0                    Style changes and automoderator                  []   
1  How do your companies keep track of multiple s...                  []   
2             How to learn programming fundamentals?                  []   
3  On house building and software development pro...                  []   
4                                   Learning to code                  []   

  subreddit_name_prefixed  hidden  pwls link_flair_css_class  downs  \
0   r/SoftwareEngineering   False     6                 None      0   
1   r/SoftwareEngineering   False     6                 None      0   
2   r/SoftwareEngineering   False     6                 None      0   
3   r/SoftwareEngineering   False     6                 None      0   
4   r/SoftwareEngineering   False     6                 None      0   

   thumbnail_height top_awarded_type  hide_score       name  quarantine  \
0               NaN             None       False  t3_cbpgfx       False   
1               NaN             None       False  t3_inz7gf       False   
2               NaN             None       False  t3_ins3wm       False   
3              78.0             None       False  t3_innvzx       False   
4               NaN             None       False  t3_ineb6z       False   

  link_flair_text_color  upvote_ratio author_flair_background_color  \
0                  dark          0.72                          None   
1                  dark          0.84                          None   
2                  dark          0.82                          None   
3                  dark          0.70                          None   
4                  dark          0.95                          None   

  subreddit_type  ups  total_awards_received media_embed  thumbnail_width  \
0         public    3                      0          {}              NaN   
1         public    4                      0          {}              NaN   
2         public   11                      0          {}              NaN   
3         public    4                      0          {}            140.0   
4         public   14                      0          {}              NaN   

  author_flair_template_id  is_original_content user_reports secure_media  \
0                     None                False           []         None   
1                     None                False           []         None   
2                     None                False           []         None   
3                     None                False           []         None   
4                     None                False           []         None   

   is_reddit_media_domain  is_meta category secure_media_embed  \
0                   False    False     None                 {}   
1                   False    False     None                 {}   
2                   False    False     None                 {}   
3                   False    False     None                 {}   
4                   F

In [9]:
# Unique IDs/names given to each post in the subreddit
pd.DataFrame(url_posts)['name']

0     t3_cbpgfx
1     t3_inz7gf
2     t3_ins3wm
3     t3_innvzx
4     t3_ineb6z
5     t3_inhaxx
6     t3_in1l8s
7     t3_imsnpc
8     t3_imys4v
9     t3_imi669
10    t3_impuyf
11    t3_im52nv
12    t3_imd8wh
13    t3_im7334
14    t3_im9s8c
15    t3_ilul2d
16    t3_im3v63
17    t3_ilswz1
18    t3_ilsny2
19    t3_ilk9u6
20    t3_ilsogp
21    t3_ilw6nr
22    t3_ilmg3e
23    t3_il6c72
24    t3_iller3
25    t3_ilg75b
Name: name, dtype: object

In [10]:
# The after column is an indication of the name of the last post on a subreddit on a particular page
url_dict['data']['after']

't3_ilg75b'

In [11]:
# Generate the url that gives us the next page with the next 25 posts
url + '?after=' + url_dict['data']['after']

'https://www.reddit.com/r/softwareengineering.json?after=t3_ilg75b'

Based on the above actions we have donne on this URL, we are now able to create our own scraping function to perform on a SubReddit URL Page.

## Scraping Software Engineering Subreddit Data Using Reddit's API

In [12]:
# Scraping Software Engineering Subreddit Data

swe_posts=[] # Storing Software Engineering Subreddit Posts
after = None # after will be empty for the first iteration
swe_url = 'https://www.reddit.com/r/softwareengineering.json'

for a in range(40): #range number will indicate how many pages of 25 posts to scrape
    if after == None:
        current_url = swe_url
    else:
        current_url = swe_url + '?after=' + after
    print(current_url)
        
    res = requests.get(current_url, headers={'User-agent': 'GA Boy'})
    
    if res.status_code == 200: #to check if request is successful
        current_dict = res.json()
        current_posts = [p['data'] for p in current_dict['data']['children']]
        swe_posts.extend(current_posts) # Adding all posts scraped into this list
        after = current_dict['data']['after']
    else:
        print(results.status_code)
        break
        
    # generate a random sleep duration to look more 'natural'
    sleep_duration = random.randint(2,6)
    print(sleep_duration)
    time.sleep(sleep_duration)
    time.sleep(1) #seconds to sleep

https://www.reddit.com/r/softwareengineering.json
4
https://www.reddit.com/r/softwareengineering.json?after=t3_ilg75b
6
https://www.reddit.com/r/softwareengineering.json?after=t3_ijqcb3
3
https://www.reddit.com/r/softwareengineering.json?after=t3_ihkilo
2
https://www.reddit.com/r/softwareengineering.json?after=t3_ifotao
3
https://www.reddit.com/r/softwareengineering.json?after=t3_idd0wf
2
https://www.reddit.com/r/softwareengineering.json?after=t3_ib9bqs
3
https://www.reddit.com/r/softwareengineering.json?after=t3_i8n0qw
6
https://www.reddit.com/r/softwareengineering.json?after=t3_i75nw7
4
https://www.reddit.com/r/softwareengineering.json?after=t3_i4mgnc
4
https://www.reddit.com/r/softwareengineering.json?after=t3_i2ja4v
5
https://www.reddit.com/r/softwareengineering.json?after=t3_i041d0
5
https://www.reddit.com/r/softwareengineering.json?after=t3_hyhhx3
3
https://www.reddit.com/r/softwareengineering.json?after=t3_hvq26m
2
https://www.reddit.com/r/softwareengineering.json?after=t3_htr2o

In [13]:
# Check the total number of posts scrapped
print(str('The total number of subreddit posts scrapped was:'), len(swe_posts))

The total number of subreddit posts scrapped was: 994


In [14]:
# Converting these scrapped posts into a dataframe
swe_df = pd.DataFrame(swe_posts)
swe_df.head()

approved_at_utc            subreddit  \
0            None  SoftwareEngineering   
1            None  SoftwareEngineering   
2            None  SoftwareEngineering   
3            None  SoftwareEngineering   
4            None  SoftwareEngineering   

                                            selftext author_fullname  saved  \
0  Hi /r/SoftwareEngineering - This is just to le...        t2_6w5wq  False   
1  I have been at several large companies now and...      t2_59dwm06  False   
2  I am doing a Uni course on Data Engineering, w...     t2_57j4x5fp  False   
3                                                           t2_ayl6m  False   
4  I’ve been learning to code using Udemy and thi...     t2_60gqusa6  False   

  mod_reason_title  gilded  clicked  \
0             None       0    False   
1             None       0    False   
2             None       0    False   
3             None       0    False   
4             None       0    False   

                                               title link_flair_richtext  \
0                    Style changes and automoderator                  []   
1  How do your companies keep track of multiple s...                  []   
2             How to learn programming fundamentals?                  []   
3  On house building and software development pro...                  []   
4                                   Learning to code                  []   

  subreddit_name_prefixed  hidden  pwls link_flair_css_class  downs  \
0   r/SoftwareEngineering   False     6                 None      0   
1   r/SoftwareEngineering   False     6                 None      0   
2   r/SoftwareEngineering   False     6                 None      0   
3   r/SoftwareEngineering   False     6                 None      0   
4   r/SoftwareEngineering   False     6                 None      0   

   thumbnail_height top_awarded_type  hide_score       name  quarantine  \
0               NaN             None       False  t3_cbpgfx       False   
1               NaN             None       False  t3_inz7gf       False   
2               NaN             None       False  t3_ins3wm       False   
3              78.0             None       False  t3_innvzx       False   
4               NaN             None       False  t3_ineb6z       False   

  link_flair_text_color  upvote_ratio author_flair_background_color  \
0                  dark          0.72                          None   
1                  dark          0.72                          None   
2                  dark          0.76                          None   
3                  dark          0.58                          None   
4                  dark          1.00                          None   

  subreddit_type  ups  total_awards_received media_embed  thumbnail_width  \
0         public    3                      0          {}              NaN   
1         public    3                      0          {}              NaN   
2         public    8                      0          {}              NaN   
3         public    2                      0          {}            140.0   
4         public   15                      0          {}              NaN   

  author_flair_template_id  is_original_content user_reports secure_media  \
0                     None                False           []         None   
1                     None                False           []         None   
2                     None                False           []         None   
3                     None                False           []         None   
4                     None                False           []         None   

   is_reddit_media_domain  is_meta category secure_media_embed  \
0                   False    False     None                 {}   
1                   False    False     None                 {}   
2                   False    False     None                 {}   
3                   False    False     None                 {}   
4                   F

In [15]:
swe_df.shape

(994, 111)

To go one step deeper, let's find the number of posts that have a unique entry in the 'selftext' column. We will remove all duplicate rows.

In [16]:
# Check the number of posts scrapped that have a unique entry in the 'selftext' column
print(str('The number of subreddit posts that have a unique entry in the "selftext" column was:'), len(set([p['selftext'] for p in swe_posts])))

The number of subreddit posts that have a unique entry in the "selftext" column was: 724


In [17]:
# Check on duplicate rows/posts
swe_df.duplicated(subset=['selftext']).sum()

270

In [18]:
# Removing duplicate rows
swe_df.drop_duplicates(subset = ['selftext'], inplace = True)

In [19]:
# Verification Check
swe_df.duplicated(subset=['selftext']).sum()

0

Secondly, we will also remove dupliicate posts that do not have a unique entry in the 'title' column.

In [20]:
# Check on duplicate rows/posts
swe_df.duplicated(subset=['title']).sum()

3

In [21]:
# Removing duplicate rows
swe_df.drop_duplicates(subset = ['title'], inplace = True)

In [22]:
# Verification Check
swe_df.duplicated(subset=['title']).sum()

0

In [23]:
# Final Remaining Rows in our swe_df
swe_df.shape

(721, 111)

After removing our duplicate rows, we are left with 721 subreddit posts for our software engineering subreddit thread.

## Scraping Data Science Subreddit Data using Reddit's API

In [24]:
# Scraping Data Science Subreddit Data

ds_posts=[] # Storing Data Science Subreddit Posts
after = None # Param will be empty for the first iteration
ds_url = 'https://www.reddit.com/r/datascience.json'

for a in range(40): #range number will indicate how many pages of 25 posts to scrape
    if after == None:
        current_url = ds_url
    else:
        current_url = ds_url + '?after=' + after
    print(current_url)
        
    res = requests.get(current_url, headers={'User-agent': 'GA Boy'})
    
    if res.status_code==200: #to check if request successful
        current_dict = res.json()
        current_posts = [p['data'] for p in current_dict['data']['children']]
        ds_posts.extend(current_posts) 
        after = current_dict['data']['after']
    else:
        print(results.status_code)
        break
        
    # generate a random sleep duration to look more 'natural'
    sleep_duration = random.randint(2,6)
    print(sleep_duration)
    time.sleep(sleep_duration)
    time.sleep(1) #seconds to sleep

https://www.reddit.com/r/datascience.json
3
https://www.reddit.com/r/datascience.json?after=t3_in5vde
6
https://www.reddit.com/r/datascience.json?after=t3_il6rux
4
https://www.reddit.com/r/datascience.json?after=t3_ii0idb
3
https://www.reddit.com/r/datascience.json?after=t3_ih1sd5
2
https://www.reddit.com/r/datascience.json?after=t3_igewo5
6
https://www.reddit.com/r/datascience.json?after=t3_iflow4
5
https://www.reddit.com/r/datascience.json?after=t3_ie7rzz
3
https://www.reddit.com/r/datascience.json?after=t3_idu7pn
6
https://www.reddit.com/r/datascience.json?after=t3_icsad1
6
https://www.reddit.com/r/datascience.json?after=t3_ib1enn
2
https://www.reddit.com/r/datascience.json?after=t3_i98wum
5
https://www.reddit.com/r/datascience.json?after=t3_i87f3l
3
https://www.reddit.com/r/datascience.json?after=t3_i639kx
6
https://www.reddit.com/r/datascience.json?after=t3_i5cwxj
6
https://www.reddit.com/r/datascience.json?after=t3_i2jogq
6
https://www.reddit.com/r/datascience.json?after=t3_i1yfl

In [25]:
# Check the total number of posts scrapped
print(str('The total number of subreddit posts scrapped was:'), len(ds_posts))

The total number of subreddit posts scrapped was: 995


In [26]:
# Converting these scrapped posts into a dataframe
ds_df = pd.DataFrame(ds_posts)
ds_df.head()

approved_at_utc    subreddit  \
0            None  datascience   
1            None  datascience   
2            None  datascience   
3            None  datascience   
4            None  datascience   

                                            selftext author_fullname  saved  \
0  Welcome to this week's entering &amp; transiti...     t2_4l4cxw07  False   
1  I wrote this post 6 months ago, and I know tha...     t2_3pmyfxcf  False   
2  Does anyone have thoughts on Chartered Statist...     t2_6324ybwh  False   
3  I'm really interested in GIS and environmental...     t2_2448korq  False   
4  Goal: Predict the amount of open bug tickets d...     t2_5j5ny4xv  False   

  mod_reason_title  gilded  clicked  \
0             None       0    False   
1             None       0    False   
2             None       0    False   
3             None       0    False   
4             None       0    False   

                                               title link_flair_richtext  \
0  Weekly Entering &amp; Transitioning Thread | 0...                  []   
1                         What we look for in hiring                  []   
2  What professional accreditation and certificat...                  []   
3  Any data scientist work with geography, GIS, E...                  []   
4                  Bug prediction with AI | Question                  []   

  subreddit_name_prefixed  hidden  pwls link_flair_css_class  downs  \
0           r/datascience   False     6                 None      0   
1           r/datascience   False     6               career      0   
2           r/datascience   False     6           discussion      0   
3           r/datascience   False     6               career      0   
4           r/datascience   False     6           discussion      0   

   thumbnail_height top_awarded_type  hide_score       name  quarantine  \
0               NaN             None       False  t3_inkv4p       False   
1               NaN             None       False  t3_innazn       False   
2               NaN             None       False  t3_io4qai       False   
3               NaN             None       False  t3_inwz0c       False   
4               NaN             None        True  t3_io5uxk       False   

  link_flair_text_color  upvote_ratio author_flair_background_color  \
0                  dark          1.00                                 
1                  dark          0.97                          None   
2                  dark          0.83                          None   
3                  dark          0.92                          None   
4                  dark          1.00                          None   

  subreddit_type  ups  total_awards_received media_embed  thumbnail_width  \
0         public    1                      0          {}              NaN   
1         public  671                      1          {}              NaN   
2         public    4                      0          {}              NaN   
3         public    9                      0          {}              NaN   
4         public    1                      0          {}              NaN   

  author_flair_template_id  is_original_content user_reports secure_media  \
0                     None                False           []         None   
1                     None                False           []         None   
2                     None                False           []         None   
3                     None                False           []         None   
4                     None                False           []         None   

   is_reddit_media_domain  is_meta category secure_media_embed  \
0                   False    False     None                 {}   
1                   False    False     None                 {}   
2                   False    False     None                 {}   
3                   False    False     None                 {}   
4                   False    False     None                 {}   

  

In [27]:
# Check on duplicate rows/posts
ds_df.duplicated(subset=['selftext']).sum()

433

In [28]:
# Removing duplicate rows
ds_df.drop_duplicates(subset = ['selftext'], inplace = True)

In [29]:
# Verification Check
ds_df.duplicated(subset=['selftext']).sum()

0

In [30]:
# Check on duplicate rows/posts
ds_df.duplicated(subset=['title']).sum()

0

In [31]:
# Final Remaining Rows in our ds_df
ds_df.shape

(562, 112)

After removing our duplicate rows, we are left with 562 subreddit posts for our data science subreddit thread.

## Saving our Data into CSV Files

In [32]:
# Saving swe_df into a csv_file
swe_df.to_csv('./data/swe_data.csv', index = False)

In [33]:
# Saving ds_df into a csv_file
ds_df.to_csv('./data/ds_data.csv', index = False)